src/metatags/opportunityHandler.js

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { TITLE, DESCRIPTION, H1 } from './constants.js';
+
+function removeTrailingSlash(url) {
+  return url.endsWith('/') ? url.slice(0, -1) : url;
+}
+
+/**
+ * Synchronizes existing suggestions with new data
+ * by removing outdated suggestions and adding new ones.
+ *
+ * @param {Object} params - The parameters for the sync operation.
+ * @param {Object} params.opportunity - The opportunity object to synchronize suggestions for.
+ * @param {Array} params.newData - Array of new data objects to sync.
+ * @param {Function} params.buildKey - Function to generate a unique key for each item.
+ * @param {Function} params.mapNewSuggestion - Function to map new data to suggestion objects.
+ * @param {Object} params.log - Logger object for error reporting.
+ * @returns {Promise<void>} - Resolves when the synchronization is complete.
+ */
+export async function syncSuggestions({
+  opportunity,
+  newData,
+  buildKey,
+  mapNewSuggestion,
+  log,
+}) {
+  const existingSuggestions = await opportunity.getSuggestions();
+  const existingSuggestionsMap = new Map(
+    existingSuggestions.map((existing) => [buildKey(existing), existing]),
+  );
+
+  // Create new suggestions and sync them with existing suggestions
+  const newSuggestions = newData
+    // map new audit data to suggestions format
+    .map(mapNewSuggestion)
+    // update new suggestions with data from existing suggestion of same key
+    .map((newSuggestion) => {
+      const existing = existingSuggestionsMap.get(buildKey(newSuggestion));
+      if (existing) {
+        return {
+          ...newSuggestion,
+          status: existing.getStatus(),
+          data: {
+            ...newSuggestion.data,
+            ...(existing.data.aiSuggestion && { aiSuggestion: existing.data.aiSuggestion }),
+            ...(existing.data.aiRationale && { aiRationale: existing.data.aiRationale }),
+          },
```

Comment:
1. how about the case where a ESS UI user has edited the suggestion and provided a custom one ("data.toOverride")

We should retain that property as well.

2. Also, Can you add a TODO comment in the code that once the Suggestions Entity supports createdBy field, we should not delete those existing suggestion entries.

Reply:
> how about the case where a ESS UI user has edited the suggestion and provided a custom one ("data.toOverride"). We should retain that property as well.

Sure, makes sense if we are thinking of using the override field for prioritisation over BO UI edits and llm training.

Added TODO comment as well.